Sha ranges do not get linkified correctly

brumm opened the pull request:
Hey Titus!

I recently came across a thing that remark-github fails to linkify correctly: comparing changes across commits, like here: https://github.com/facebook/react/pull/21480#issuecomment-836376580

Instead of turning two commit SHAs combined by `...` into a comparison link, only the first SHA is linkified like so: 
![image](https://user-images.githubusercontent.com/170500/117671104-be48db80-b1a8-11eb-90cd-3771152a1416.png)

I added a failing test and a fix for this. I wasn't sure whether the fixture I chose is a good place to add the test, please let me know if you'd like me to change something.

wooorm commented:
```markdown
Well, would you look at that: e2acebc...2aa9311

Actual unicode ellipsis: e2acebc…2aa9311

punctuation before: -e2acebc...2aa9311

e2acebc...2aa9311.punctuation after

two: e2acebc..2aa9311

one: e2acebc.2aa9311

fork? brumm@e2acebc...2aa9311

repo: remarkjs/remark-github@e2acebc...2aa9311

Pasting the link: https://github.com/remarkjs/remark-github/compare/e2acebcb40ae1bbe5b3ed87488f294faa45b2b85...2aa931154d6e0f6317aff644924163999cb61894
```

Well, would you look at that: e2acebc...2aa9311

Actual unicode ellipsis: e2acebc…2aa9311

punctuation before: -e2acebc...2aa9311

e2acebc...2aa9311.punctuation after

two: e2acebc..2aa9311

one: e2acebc.2aa9311

fork? brumm@e2acebc...2aa9311

repo: remarkjs/remark-github@e2acebc...2aa9311

Pasting the link: https://github.com/remarkjs/remark-github/compare/e2acebcb40ae1bbe5b3ed87488f294faa45b2b85...2aa931154d6e0f6317aff644924163999cb61894

wooorm commented:
Some more:

```markdown
5: e2aceb...2aa931

7/5: e2acebc...2aa931

5/7: e2aceb...2aa9311

e2acebc...

...2aa9311
```

5: e2aceb...2aa931

7/5: e2acebc...2aa931

5/7: e2aceb...2aa9311

e2acebc...

...2aa9311

---

It would probably be a good idea to add some of these to a new fixture file, to ensure we match the behavior correctly. Interested in that? Reverse engineering GH is annoying and time consuming tho 😅

brumm commented:
While working on this, I changed the test setup to do line-based equality checks, which results in a much nicer dev experience. I also added a TAP formatter. Would you be interested in another PR for this?

wooorm commented:
I also use tap formatters myself, but they can be chained after, like so: `node test | tap-difflet`.
Given that I don’t always want diffs, I think it’s better to keep it without?

For the per-line tests, I’m :+1: to that!

wooorm commented on test/fixtures/sha-range/input.md:
```diff
@@ -0,0 +1,27 @@
+
+repo: remarkjs/remark-github@e2acebc...2aa9311
+
+Pasting the link: https://github.com/remarkjs/remark-github/compare/e2acebcb40ae1bbe5b3ed87488f294faa45b2b85...2aa931154d6e0f6317aff644924163999cb61894
```
that might be a micromark-extension-gfm-... bug then, weird!

wooorm replied:
Wait, remark-gfm isn't used in the tests? 🤔😅

wooorm replied:
Well that explains why literal URLs are not seen as URLs. Could you add that to the tests? And hopefully everything will keep working

brumm replied:
that'd be great!

wooorm commented:
What’s failing in the test? Is it the full pasted in url? If so that can be solved by handling it [here](https://github.com/remarkjs/remark-github/blob/f2fed6a1695ebf89fbff8f898f126f97e60609f7/index.js#L256). Or something else?

brumm commented:
Hey, life and work happened. I'll take this on either tonight or tomorrow.

brumm commented:
> What’s failing in the test? Is it the full pasted in url? If so that can be solved by handling it [here](https://github.com/remarkjs/remark-github/blob/f2fed6a1695ebf89fbff8f898f126f97e60609f7/index.js#L256). Or something else?

After adding remark-gfm two previously green tests are failing:

- `brumm@e2acebc...2aa9311` gets turned into `<brumm@e2acebc...2aa9311>`
- `remarkjs/remark-github@e2acebc...2aa9311` gets turned into `remarkjs/remark-<github@e2acebc...2aa9311>`

This prevents the replacers in `findAndReplace` from matching. 

The pasted link is also affected by remark-gfm:

`https://github.com/remarkjs/remark-github/compare/e2acebcb40ae1bbe5b3ed87488f294faa45b2b85...2aa931154d6e0f6317aff644924163999cb61894` gets turned into `<https://github.com/remarkjs/remark-github/compare/e2acebcb40ae1bbe5b3ed87488f294faa45b2b85...2aa931154d6e0f6317aff644924163999cb61894>`

Am I assuming correctly that this is a problem with remark-gfm's linkification? Maybe stemming from micromark-extension-gfm-autolink-literal? I'm a bit lost :)

wooorm commented:
For 1) that’s because I think you’re not handling such projects for commit ranges yet, whereas they are handled for [references](https://github.com/remarkjs/remark-github/blob/62f70ca375c0a54087ce0bd8e08ea035ed2d96c4/index.js#L67-L71)?

The 2), the pasted link, it’s indeed supposed to be turned into an actual link first. And *then* we need code to handle it: normal links should be visited, [parsed](https://github.com/remarkjs/remark-github/blob/62f70ca375c0a54087ce0bd8e08ea035ed2d96c4/index.js#L285), and then the [contents of the link rewritten](https://github.com/remarkjs/remark-github/blob/62f70ca375c0a54087ce0bd8e08ea035ed2d96c4/index.js#L248) I think.

brumm commented:
Alright, I got a shoddy fix for the 'pasted code' test. Doesn't seem to be the greatest way of doing this, but I'm running a bit out of time on this :)

I think you might be on to something about the other bits getting turned into emails, but unfortunately that's a rabbit hole I won't be able to go down :(

wooorm commented:
Thanks for your patience, released!

brumm commented:
Thank _you_ so much for picking up my slack!
